### PR TITLE
feat(elasticloadbalancingv2): add load balancer lookups

### LINKED ISSUE #11088: [elasticloadbalancingv2] load balancers by lookup
I'd like my constructs to discover load balancers and listeners by tags, ports, and protocols.

### Use Case

We share load balancers on shared ECS clusters between several CDK-based services for cost-reduction reasons. Because of this, we often need to import the load balancer and listener configuration. Right now we're using `.from*Attributes()` method, but this results in a lot of copy-pasting of attribute values. If we could look up a load balancer or listener by tags, then we could replace the copy-pasting with a tag-based discovery method.

### Proposed Solution

We could implement a context provider for load balancers and allow the user to look up values the same way as `Vpc.fromLookup()`.

### Other

I'm working on a PR for this right now. I'll mention this issue when I've got the PR up.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by misterjoshua**
Adds `fromLookup()` methods to both Application and Network load balancers as well as their listeners.

Closes #11088 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by misterjoshua** [comment]
I've looked over my code since I put this PR up. Given that this change entails a new manifest schema version, and that I have implemented looking up security groups by id: Is now a good time to add tag search for security groups?

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts:111]:
```suggestion
    let listenerProtocol: cxschema.LoadBalancerListenerProtocol | undefined;
```

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts:131]: This seems like a case where following a pattern and having commonality between the ALB and NLB interface yields a sub-optimal result. `NetworkListenerLookupOptions` extends `BaseListenerLookupOptions`, which includes `listenerArn` as one of the filtering options, but there's no case here for why someone would use `fromLookup` with `listenerArn` over `fromNetworkListenerArn`.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener.ts:578]: I think a cleaner approach would be to keep `connections` and `listenerArn` abstract in this base class, and just set them in the concrete classes. Saves needing to create the `ImportedApplicationListenerBaseProps` and awkward `super` calls.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener.ts:135]:
```suggestion
      listenerProtocol: options.listenerProtocol === ApplicationProtocol.HTTP
```

Please use `===` and `!==` throughout.

**Review by njlynch — CHANGES_REQUESTED** [review]
Thanks for the contribution! I would have preferred slightly smaller chunks over something this size -- e.g., just the load balancers by ARNs/tags would have been significant in itself. That being said, we'll take this as-is. I did a first pass of comments, will go deeper on the next round.

**Comment by misterjoshua** [comment]
Thanks for the review. I'll get to these changes as soon as possible. :)

File: packages/@aws-cdk/aws-ec2/lib/security-group.ts
Comment (misterjoshua) [reply packages/@aws-cdk/aws-ec2/lib/security-group.ts:317]: Only because I found something similar in `Vpc.fromLookup()`, where the dummy value is held separately and not given as `dummyValue` input to `getValue`. Given that I don't have strong feelings here, I'll change this to just use `dummyValue` flat out.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts
Comment (misterjoshua) [reply packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts:131]: @njlynch Makes sense. Is your suggestion then to entirely separate the interfaces for `ApplicationListener.fromLookup()` and `NetworkListener.fromLookup()`? Or to push `BaseListenerLookupOptions.listenerArn` down to `ApplicationListenerLookupOptions.listenerArn`? Or...?

Comment (njlynch) [reply packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts:131]: Well, as a bit of devil's advocate, what use case does adding the lookup by ARN enable? What specific functionality for `ApplicationListener.fromLookup()` is benefitted from `ApplicationListener.fromLookup({ listenerArn: 'foo' })` vs `ApplicationListener.fromApplicationListenerAttributes()` with the ARN?

File: packages/@aws-cdk/aws-ec2/lib/security-group.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-ec2/lib/security-group.ts:317]: The VPC is the odd one out. There was a specific subnet-related problem there that needed to be special-cased (see #4090) .

File: packages/aws-cdk/lib/context-providers/load-balancers.ts
Comment (misterjoshua) [reply packages/aws-cdk/lib/context-providers/load-balancers.ts:180]: Whoops. Good catch!

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts
Comment (misterjoshua) [reply packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts:131]: @njlynch Fair enough comments. But I had in mind that if the user has the ARN of the listener, it's nice to have CDK get the security groups of the ALB by lookup as well. For instance, the user is probably going to need the ALB security groups to let the ALB traffic into an ECS/Fargate task.

They can certainly use `ApplicationListener.fromApplicationListenerAttributes()` with the ids of all involved resources, but then they'll need to provide the security groups of the load balancer as well. So, the improvement of lookup by ARN is primarily a convenience over needing to specify the security group.

If NLB's listener's interface had `ec2.IConnectable`, I'd probably say the same for it, but only `IApplicationListener` extends the interface. So ALB and NLB listeners are definitely a little incongruous in that respect.

File: packages/aws-cdk/lib/context-providers/load-balancers.ts
Comment (misterjoshua) [reply packages/aws-cdk/lib/context-providers/load-balancers.ts:230]: @njlynch Hmm. I see. If I change the code as follows, I can handle an undefined `Value` and the loop conditional will work with strict not equals.

```ts
const tagsByName: Record<string, string | undefined> = {};
for (const tag of tagDescription.Tags ?? []) {
  tagsByName[tag.Key!] = tag.Value;
}
```

However, unless we change `cxschema.Tag`, we can't match on an undefined tag value. From a quick glance, this change would entail a few changes to `CloudFormationStackArtifact`. What do you think we should do?

File: packages/@aws-cdk/aws-ec2/lib/security-group.ts
Comment (misterjoshua) [reply packages/@aws-cdk/aws-ec2/lib/security-group.ts:302]: This is in.

File: packages/aws-cdk/lib/context-providers/load-balancers.ts
Comment (misterjoshua) [reply packages/aws-cdk/lib/context-providers/load-balancers.ts:164]: I've implemented these checks higher up because the context providers know better.

**Comment by misterjoshua** [comment]
@njlynch Hey. I think I've got most of your review items addressed in the latest batch of commits. There are two outstanding things as far as I can tell:

* The use case for listener lookups by ARN. (My response is in the review thread.)
* Need to figure out what to do about `cxschema.Tag.value` not allowing undefined. (Comments in the review thread.)

In the meantime, I'll keep an eye out for any more feedback.

File: packages/aws-cdk/lib/context-providers/load-balancers.ts
Comment (misterjoshua) [reply packages/aws-cdk/lib/context-providers/load-balancers.ts]: This behavior is now updated.

**Comment by misterjoshua** [comment]
After the merge, the build broke. It looks like I may need to add another allowed breaking change. JSII 1.14 notices the three new query types in a new way:

```
[weakened:@aws-cdk/cloud-assembly-schema.MissingContext]
```

From the output, jsii seems to indicate that the new `MissingContext` can't be assigned to the old `MissingContext`. This is to be expected, considering that we've added three arms to the context queries union type and have updated the schema major version.

I see that 1.71.0 is in the wings, so I'll wait to merge that in before I push up another build.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts:131]: Fair point. I can see the benefit of having the security groups automatically loaded. Perhaps we move `listenerArn` up to `ApplicationListenerLookupOptions` then. I'm also going to ask one of my colleagues to take a look as this PR since it's so big, and see if they have an opinion.

File: allowed-breaking-changes.txt
Comment (misterjoshua) [inline allowed-breaking-changes.txt:58]: Changing line 58 to the following fixes the problem with this PR's build, introduced when merging in JSII 1.14 from master:

```
weakened:@aws-cdk/cloud-assembly-schema.MissingContext
````

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.ts:131]: I like moving `listenerArn` to `ApplicationListenerLookupOptions` if providing the ARN doesn't make sense for Network Listeners.

We can always move it back down, but we can't remove it after it's added to `NetworkListenerLookupOptions` 🙂.

File: packages/@aws-cdk/cx-api/lib/context/security-group.ts
Comment (skinny85) [inline packages/@aws-cdk/cx-api/lib/context/security-group.ts:12]: `SecurityGroup.fromLookup()` can also be called without a LoadBalancer context... what does this property mean then?

**Review by skinny85 — COMMENTED** [review]
Looks fine to me. Couple of minor issues that I've found.

Since this is such a complicated change, I would loop in @rix0rrr to sanity-check only the context provider changes look good, and @iliapolo just to verify the cx-api bump. Thanks guys!

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener.ts
Comment (misterjoshua) [reply packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener.ts:135]: Bug. My bad. I'll fix. I intended that it should do as the docs say.

File: packages/@aws-cdk/cx-api/lib/context/security-group.ts
Comment (misterjoshua) [reply packages/@aws-cdk/cx-api/lib/context/security-group.ts:12]: It's more general-purpose than this comment conveys. I'll adjust the wording so that this is clear.

**Comment by misterjoshua** [comment]
@skinny85 I believe I've addressed your review items in the latest commits. I also updated from the main branch. Everything seems to be in working order. I'll continue to keep an eye out for more changes. :)

**Comment by iliapolo** [comment]
@skinny85 @misterjoshua I'm taking a look at the cx-api changes, added `do-not-merge` in the meantime just in case.

**Comment by skinny85** [comment]
Thanks @iliapolo !

@rix0rrr if you could take a quick look at just the provider changes, that would be great.

**Comment by misterjoshua** [comment]
While you guys are reviewing, would you prefer that I resolve the merge conflicts or handle them for later?

**Comment by njlynch** [comment]
> While you guys are reviewing, would you prefer that I resolve the merge conflicts now or handle them later?

Feel free to resolve them now.

**Comment by iliapolo** [comment]
@njlynch @skinny85 I'm good with the cx-api changes, so removing the `do-not-merge` I added.

File: packages/aws-cdk/lib/context-providers/load-balancers.ts
Comment (njlynch) [inline packages/aws-cdk/lib/context-providers/load-balancers.ts:73]: `~` (Style) I'd probably just inline this:

```suggestion
    return query.listenerArn ? this.getListenerByArn(elbv2, query) : this.getListenerByFilteringLoadBalancers(elbv2, query);
```

**Review by njlynch — COMMENTED** [review]
Really close, just one legitimate comment and a few stylistics nitpicks.

File: packages/@aws-cdk/aws-elasticloadbalancingv2/lib/shared/base-listener.ts
Comment (misterjoshua) [reply packages/@aws-cdk/aws-elasticloadbalancingv2/lib/shared/base-listener.ts:89]: The props are a `{ [key: string]: any }`, so the cast isn't strictly necessary. But, I think the type hint is valuable if only for the reader's understanding, as otherwise, the correct type for the context query isn't obvious on inspection. I'll defer to your judgement on this matter though. :)

**Review by njlynch — APPROVED** [review]
Great! Thanks for the contribution!

**Pull request merged**
